### PR TITLE
[gptq] add gptq kernel

### PR DESCRIPTION
**Pull request by Xu-Kai**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/gptq/cai_gptq/cai_quant_linear.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        # self.register_buffer('input_idx', torch.zeros(infeatures], dtype=torch.int32))
+
+
+        if bias:
+            self.register_buffer('bias', torch.zeros((outfeatures), dtype=torch.float16))
+        else:
+            self.bias = None
+
+        self.gptq_linear = CaiGPTQLinearOp(groupsize, bits)
+        self.printed = False
```
Comment (ver217) [inline colossalai/gptq/cai_gptq/cai_quant_linear.py:36]: It seems this attr is not used. Can we delete it?

File: colossalai/gptq/gptq_utils/gptq.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+import transformers
+from .quant import Quantizer
+from texttable import Texttable
+from .utils import torch_snr_error
+
+torch.backends.cuda.matmul.allow_tf32 = False
+torch.backends.cudnn.allow_tf32 = False
+
+
+class Observer:
```
Comment (ver217) [inline colossalai/gptq/gptq_utils/gptq.py:15]: What is the usage of this class?

File: colossalai/gptq/gptq_utils/gptq.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+import time
+
+import torch
+import torch.nn as nn
+import transformers
+from .quant import Quantizer
+from texttable import Texttable
+from .utils import torch_snr_error
+
+torch.backends.cuda.matmul.allow_tf32 = False
```
Comment (ver217) [inline colossalai/gptq/gptq_utils/gptq.py:11]: Why disable tf32?

File: colossalai/gptq/gptq_utils/gptq.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
+import math
+import time
+
+import torch
```
Comment (ver217) [inline colossalai/gptq/gptq_utils/gptq.py:4]: Is the `gptq_utils` copied from GPTQ? If so, we should claim this in all copied files and add their license to ours.

File: tests/test_gptq/linear_act_fusion_bench.py [hunk 5]
```diff
@@ -0,0 +1,383 @@
+
```
Comment (ver217) [inline tests/test_gptq/linear_act_fusion_bench.py:1]: Benchmark should not be put in `tests` folder

Comment (ver217) [reply tests/test_gptq/linear_act_fusion_bench.py:1]: You could put it in `examples`

File: tests/test_gptq/test_linear_act_fusion.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
+
```
Comment (ver217) [inline tests/test_gptq/test_linear_act_fusion.py:1]: You should add `text_xxx()` functions as test entry.

File: examples/ops/gptq/linear_act_fusion_bench.py [hunk 7]
```diff
@@ -0,0 +1,382 @@
+import torch
```
Comment (ver217) [inline examples/ops/gptq/linear_act_fusion_bench.py:1]: Can you rename the folder to `inference/quant`?

File: requirements/requirements.txt [hunk 8]
```diff
@@ -12,3 +12,5 @@ torch>=1.11
 safetensors
 flash_attn>=2.0
 einops
+texttable
+auto-gptq
```
Comment (ver217) [inline requirements/requirements.txt:16]: Can it be an optional requirements? When users want to use it, we can check whether we can import it correctly, if not we can raise an error.

File: tests/test_gptq/test_linear_act_fusion.py [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+            qweight, qscales, qzeros = cai_linear_pack(layers[name], scale, zero, 
+                        qweight, qscales, qzeros, g_idx, 
+                        layers[name].weight.shape[-1], groupsize, wbits)
+
+    # print("cai pack", layers)
+    return qweight, qscales, qzeros 
+
+
+def test_gptq_linear():
+    parser = ArgumentParser()
```
Comment (ver217) [inline tests/test_gptq/test_linear_act_fusion.py:197]: It should be have argument parser. Instead, you should use pytest parametrize

File: tests/test_gptq/test_linear_act_fusion.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+    mean_diff = torch.mean(torch.abs(torch_out - cai_out))
+    max_diff = torch.max(torch.abs(torch_out - cai_out))
+    print("torch vs cai: mean_diff=%.8f, max_diff=%.8f" % (mean_diff, max_diff))
+
+    mean_diff = torch.mean(torch.abs(batch_cai_out - batch_gptq_out))
+    max_diff = torch.max(torch.abs(batch_cai_out - batch_gptq_out))
+    print("batch cai vs gptq: mean_diff=%.8f, max_diff=%.8f" % (mean_diff, max_diff))
+    mean_diff = torch.mean(torch.abs(batch_torch_out - batch_gptq_out))
+    max_diff = torch.max(torch.abs(batch_torch_out - batch_gptq_out))
+    print("batch torch vs gptq: mean_diff=%.8f, max_diff=%.8f" % (mean_diff, max_diff))
```
Comment (ver217) [inline tests/test_gptq/test_linear_act_fusion.py:312]: You should not print it. Instead, you should make an assertation here.

File: requirements/requirements.txt (hunk 8 above)
Comment (ver217) [reply requirements/requirements.txt:16]: FYI: https://github.com/hpcaitech/ColossalAI/blob/d20dceb9a3d1bdcb2376201220f49fca7c7c1be9/colossalai/kernel/cuda_native/mha/mha.py#L13

File: colossalai/gptq/__init__.py [hunk 11]
```diff
@@ -0,0 +1,14 @@
+HAS_AUTO_GPTQ = False
+
+try:
+    import auto_gptq
+    HAS_AUTO_GPTQ = True
+except ImportError:
+    warnings.warn('please install auto-gptq from https://github.com/PanQiWei/AutoGPTQ')
+    HAS_AUTO_GPTQ = False
```
Comment (ver217) [inline colossalai/gptq/__init__.py:8]: Maybe you don't need to check it twice, and you can just import `HAS_AUTO_GPTQ` here.

**Review by ver217 — APPROVED** [review]

**Comment by Xu-Kai** [comment]
image

**Pull request merged**